File: flink_scala/src/main/scala/github/yahuili1128/wordcount/SocketWindowWordCount.scala
package github.yahuili1128.wordcount

import org.apache.flink.api.common.functions.ReduceFunction
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.api.windowing.time.Time

/**
  * Author:  LiYahui
  * Date:  Created in  2019/7/13 16:23
  * Description: TODO  flink 读取socket
  * Version: V1.0         
  */
object SocketWindowWordCount {
  def main(args: Array[String]): Unit = {

    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    import org.apache.flink.streaming.api.scala._
    val socketData: DataStream[String] = env.socketTextStream("liyahui-01", 9999, '\n')
    val wordcountDS: DataStream[(String, Int)] = socketData.flatMap(_.split("\\s"))
        .map((_, 1))
        .keyBy(_._1)
        .timeWindow(Time.seconds(5), Time.seconds(1))
        .reduce(new ReduceFunction[(String, Int)] {
          override def reduce(t: (String, Int), t1: (String, Int)): (String, Int) = {
            (t._1, t._2 + t1._2)
          }
        })

    //    reduce 的作用可以用sum替代

    wordcountDS.print()
        .setParallelism(1)
    env.execute(s"${this.getClass.getSimpleName}")
  }
}

File: flink_java/src/main/java/github/yahuili1128/wordcount/SocketWindowWordCount.java
package github.yahuili1128.wordcount;

import github.yahuili1128.pojo.WordCount;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.functions.ReduceFunction;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.util.Collector;

/**
 * @Description : java flink socket wordcount
 * @Author : LiYahui
 * @Date : 2019-06-25 16:37
 * @Version : V1.0
 */
public class SocketWindowWordCount {
	public static void main(String[] args) throws Exception {
		//	the port to connect to
		final int port;
		try {
			ParameterTool parameterTool = ParameterTool.fromArgs(args);
			port = parameterTool.getInt("port");
		} catch (Exception e) {
			System.out.println("No port specified. Please run 'SocketWindowWordCount --port <port>'");
			return;
		}
		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

		DataStreamSource<String> text = env.socketTextStream("liyahui-01", port, "\n");


		SingleOutputStreamOperator<WordCount> wordcounts = text.flatMap(new FlatMapFunction<String, WordCount>() {
			@Override
			public void flatMap(String value, Collector<WordCount> out) throws Exception {
				for (String word : value.split("\\s")) {
					out.collect(new WordCount(word, 1L));
				}
			}
		}).keyBy("word").timeWindow(Time.seconds(5), Time.seconds(1)).reduce(new ReduceFunction<WordCount>() {
			@Override
			public WordCount reduce(WordCount value1, WordCount value2) throws Exception {
				return new WordCount(value1.word, value1.count + value2.count);
			}
		});
		wordcounts.print().setParallelism(1);

		env.execute(SocketWindowWordCount.class.getClass().getSimpleName());


	}

}
